### PR TITLE
[ESIMD] Support bfloat16 simd vector element type. 

### PR DESCRIPTION
**Pull request by kbobrovs**
Implement corresponding element type traits
 
Complementary E2E test patch: https://github.com/intel/llvm-test-suite/pull/1193

Signed-off-by: Konstantin S Bobrovsky [konstantin.s.bobrovsky@intel.com](mailto:konstantin.s.bobrovsky@intel.com)

### CODE REVIEW

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp [hunk 1]
```diff
@@ -1766,6 +1775,11 @@ size_t SYCLLowerESIMDPass::runOnFunction(Function &F,
         ToErase.push_back(CI);
         continue;
       }
+      if (Name.startswith("__esimd_bitcast")) {
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply llvm/lib/SYCLLowerIR/ESIMD/LowerESIMD.cpp]: Actually it works, thanks for pointing out.

File: sycl/include/sycl/ext/intel/esimd/detail/bfloat16_type_traits.hpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext::intel::esimd::detail {
+
+using bfloat16 = sycl::ext::oneapi::experimental::bfloat16;
+
+template <> struct element_type_traits<bfloat16> {
+  // TODO map the raw type to __bf16 once SPIRV target supports it:
+  using RawT =
+      typename std::invoke_result_t<decltype(&bfloat16::raw), bfloat16>;
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply sycl/include/sycl/ext/intel/esimd/detail/bfloat16_type_traits.hpp]: Good point as well. I reworked this to just have RawT be an unsigned integer type of the same size, given that 'raw' and other bfloat16 members will be deprecated.

**Review by v-klochkov — DISMISSED** [review]

**Comment by v-klochkov** [comment]
Looks good to me. Thank you. Please just fix the clang-format.

**Review by v-klochkov — APPROVED** [review]

**Comment by kbobrovs** [comment]
CUDA testing failure is unrelated - merging

**Pull request merged**